--- a/clean/Clean/Utils/OfflineMemory.lean
+++ b/clean/Clean/Utils/OfflineMemory.lean
@@ -898,14 +898,7 @@
   have l2_strict_sorted := MemoryAccessList.addressStrictTimestampSorted_of_AddressTimestampSorted_noTimestampDup l2 h_l2_sorted l2_notimestampdup
   have l3_strict_sorted := MemoryAccessList.addressStrictTimestampSorted_of_AddressTimestampSorted_noTimestampDup l3 h_l3_sorted l3_notimestampdup
   exact List.eq_of_perm_of_sorted l2_perm_l3 l2_strict_sorted l3_strict_sorted
-/--
-  This is the main theorem of this file.
-
-  A list of timestamp-sorted memory accesses is consistent if and only if there exists a permutation of the list
-  that is address-sorted, and such that the offline consistency check holds.
--/
-theorem MemoryAccessList.isConsistentOnline_iff_isConsistentOffline
-    (accesses : MemoryAccessList)
+theorem MemoryAccessList.isConsistentOnline_iff_isConsistentOffline (accesses : MemoryAccessList)
     (h_sorted : accesses.isTimestampSorted)
     (h_nodup : accesses.Notimestampdup) :
     MemoryAccessList.isConsistentOnline accesses h_sorted ↔
@@ -914,19 +907,15 @@
       MemoryAccessList.isConsistentOffline permuted.val permuted.property := by
   constructor
   · intro h
-    use ⟨MemoryAccessList.addressTimestampSort accesses, MemoryAccessList.addressTimestampSort_sorted accesses⟩
-    constructor
-    · simp only
-      apply MemoryAccessList.addressTimestampSort_perm
-    · simp only
-      have h' := MemoryAccessList.isConsistentOnline_iff_sorted_isConsistentOffline accesses h_sorted h_nodup
-      rw [←h']
-      assumption
+    refine ⟨⟨MemoryAccessList.addressTimestampSort accesses, MemoryAccessList.addressTimestampSort_sorted accesses⟩, ?_, ?_⟩
+    · exact MemoryAccessList.addressTimestampSort_perm accesses
+    · exact (MemoryAccessList.isConsistentOnline_iff_sorted_isConsistentOffline accesses h_sorted h_nodup).mp h
   · rintro ⟨⟨permuted, h_permuted_sorted⟩, h_perm, h_offline⟩
-    simp_all only
-    rw [List.perm_comm] at h_perm
-    have h_eq := MemoryAccessList.eq_of_perm_of_sorted h_sorted h_permuted_sorted (MemoryAccessList.addressTimestampSort_sorted accesses)
-      h_perm (by rw [List.perm_comm]; apply MemoryAccessList.addressTimestampSort_perm)
+    have h_perm' : accesses.Perm permuted := List.perm_comm.mp h_perm
+    have h_sort_perm : accesses.Perm (MemoryAccessList.addressTimestampSort accesses) :=
+      List.perm_comm.mp (MemoryAccessList.addressTimestampSort_perm accesses)
+    have h_eq : permuted = MemoryAccessList.addressTimestampSort accesses := by
+      exact MemoryAccessList.eq_of_perm_of_sorted h_sorted h_permuted_sorted (MemoryAccessList.addressTimestampSort_sorted accesses) h_perm' h_sort_perm
     simp only [h_eq] at h_offline
-    rw [←MemoryAccessList.isConsistentOnline_iff_sorted_isConsistentOffline accesses h_sorted h_nodup] at h_offline
-    assumption
+    exact (MemoryAccessList.isConsistentOnline_iff_sorted_isConsistentOffline accesses h_sorted h_nodup).mpr h_offline
+
