--- a/clean/Clean/Gadgets/Addition8/Theorems.lean
+++ b/clean/Clean/Gadgets/Addition8/Theorems.lean
@@ -75,12 +75,6 @@
       apply Nat.lt_add_one_of_le
       assumption
 
-/--
-  Soundness of the 8-bit addition circuit: assuming that the constraints and assumptions
-  are satisfied, the output is correctly the sum of the inputs and the input
-  carry modulo 256. Additionally the output carry is exactly the integer division
-  of the aforementioned sum by 256.
--/
 theorem soundness (x y out carry_in carry_out : F p):
     x.val < 256 -> y.val < 256 ->
     out.val < 256 ->
@@ -88,53 +82,39 @@
     IsBool carry_out ->
     (x + y + carry_in + -out + -(carry_out * 256) = 0) ->
     (out.val = (x.val + y.val + carry_in.val) % 256
-    ∧ carry_out.val = (x.val + y.val + carry_in.val) / 256):= by
-  intros hx hy hout carry_in_bool carry_out_bool h
-  have carry_in_bound := IsBool.val_lt_two carry_in_bool
-
+    ∧ carry_out.val = (x.val + y.val + carry_in.val) / 256) := by
+  intro hx hy hout carry_in_bool carry_out_bool h
+  have carry_in_bound : carry_in.val < 2 := IsBool.val_lt_two carry_in_bool
   rcases carry_out_bool with zero_carry | one_carry
-  -- case with zero carry
   · rw [zero_carry] at h
     simp only [zero_mul, neg_zero, add_zero] at h
-    rw [←sub_eq_add_neg] at h
     have h_spec : carry_in + x + y - out = 0 := by
-      rw [add_comm (x + y), ←add_assoc] at h
-      assumption
-
+      simpa only [sub_eq_add_neg, add_assoc, add_comm, add_left_comm] using h
     have thm := soundness_zero_carry x y out carry_in hx hy hout carry_in_bound h_spec
-    apply_fun ZMod.val at zero_carry
-
-    -- now it is just a matter of shuffling terms around
+    have hco : carry_out.val = 0 := by
+      simpa only [zero_carry, ZMod.val_zero]
     have shuffle_terms : carry_in.val + x.val + y.val = x.val + y.val + carry_in.val := by
-      zify; ring
-    rw [ZMod.val_zero, ← thm.right] at zero_carry
-    rw [shuffle_terms] at thm
-    rw [shuffle_terms] at zero_carry
+      simpa only [Nat.add_assoc, Nat.add_comm, Nat.add_left_comm]
     constructor
-    · exact thm.left
-    · exact zero_carry
-
-  -- case with one carry
+    · rw [← shuffle_terms]
+      exact thm.left
+    · rw [hco, ← shuffle_terms]
+      exact thm.right.symm
   · rw [one_carry] at h
     simp only [one_mul] at h
-    -- rw [←sub_eq_add_neg, ←sub_eq_add_neg (carry_in + x + y)] at h
     have h_spec : carry_in + x + y - out - 256 = 0 := by
-      rw [add_comm (x + y), ←add_assoc] at h
-      ring_nf at h; ring_nf
-      assumption
-
-    -- instantiate the sub-theorem
+      simpa only [sub_eq_add_neg, add_assoc, add_comm, add_left_comm] using h
     have thm := soundness_one_carry x y out carry_in hx hy hout carry_in_bound h_spec
-    apply_fun ZMod.val at one_carry
-
+    have hco : carry_out.val = 1 := by
+      simpa only [one_carry, ZMod.val_one]
     have shuffle_terms : carry_in.val + x.val + y.val = x.val + y.val + carry_in.val := by
-      zify; ring
-    rw [ZMod.val_one, ← thm.right] at one_carry
-    rw [shuffle_terms] at thm
-    rw [shuffle_terms] at one_carry
+      simpa only [Nat.add_assoc, Nat.add_comm, Nat.add_left_comm]
     constructor
-    · exact thm.left
-    · exact one_carry
+    · rw [← shuffle_terms]
+      exact thm.left
+    · rw [hco, ← shuffle_terms]
+      exact thm.right.symm
+
 
 /--
   Given the default witness generation, we show that the addition constraint is satisfied
