--- a/clean/Clean/Gadgets/Xor/Xor32.lean
+++ b/clean/Clean/Gadgets/Xor/Xor32.lean
@@ -101,21 +101,14 @@
 
 theorem completeness : Completeness (F p) elaborated Assumptions := by
   intro i0 env input_var h_env input h_input as
-  let ⟨⟨ x0_var, x1_var, x2_var, x3_var ⟩,
-       ⟨ y0_var, y1_var, y2_var, y3_var ⟩⟩ := input_var
-  let ⟨⟨ x0, x1, x2, x3 ⟩,
-       ⟨ y0, y1, y2, y3 ⟩⟩ := input
+  let ⟨⟨ x0, x1, x2, x3 ⟩, ⟨ y0, y1, y2, y3 ⟩⟩ := input
   simp only [circuit_norm, explicit_provable_type, Inputs.mk.injEq, U32.mk.injEq] at h_input
-
   simp only [Assumptions, circuit_norm, U32.Normalized] at as
-  obtain ⟨ x_bytes, y_bytes ⟩ := as
-  obtain ⟨ x0_byte, x1_byte, x2_byte, x3_byte ⟩ := x_bytes
-  obtain ⟨ y0_byte, y1_byte, y2_byte, y3_byte ⟩ := y_bytes
-
   simp only [h_input, circuit_norm, main, ByteXorTable,
     explicit_provable_type, Fin.forall_iff] at h_env ⊢
   have h_env0 : env.get i0 = ↑(ZMod.val x0 ^^^ ZMod.val y0) := by simpa using h_env 0
   simp_all [xor_val]
+
 
 def circuit : FormalCircuit (F p) Inputs U32 where
   Assumptions
