--- a/veil/Examples/StellarConsensus/SCPTheory.lean
+++ b/veil/Examples/StellarConsensus/SCPTheory.lean
@@ -88,14 +88,21 @@
 theorem intact_node_is_well_behaved : ∀ n I, intact (inst := inst) I → n ∈ I → n ∈ W := by
   intro n S h ; apply intertwined_node_is_well_behaved ; apply intact_implies_intertwined _ h
 
-theorem slice_blocks_ne : ∀ n S I, intact (inst := inst) I → n ∈ I → blocks_slices S n →
-    S ∩ I ≠ ∅ := by
-  intro n S I hI hin hblock
+theorem slice_blocks_ne (n : Node) (S I : Set Node) : intact (inst := inst) I → n ∈ I → blocks_slices S n → S ∩ I ≠ ∅ := by
+  intro hI hin hblock
   unfold blocks_slices at hblock
-  have h := hI.q_avail ; unfold quorum at h
-  simp at h ; specialize h _ hin (intact_node_is_well_behaved _ _ hI hin)
-  rcases h with ⟨Sl, hSl, h⟩ ; specialize hblock _ hSl
-  rw [Set.ne_empty_iff_exists_mem] at hblock ⊢ ; simp at hblock ⊢
-  aesop
+  have h := hI.q_avail
+  unfold quorum at h
+  have hnW : n ∈ W := intact_node_is_well_behaved n I hI hin
+  have hnIW : n ∈ I ∩ W := by
+    exact ⟨hin, hnW⟩
+  specialize h n hnIW
+  rcases h with ⟨Sl, hSl, hsub⟩
+  have hne : Sl ∩ S ≠ ∅ := hblock Sl hSl
+  rw [Set.ne_empty_iff_exists_mem] at hne ⊢
+  rcases hne with ⟨x, hx⟩
+  exact ⟨x, ⟨hx.2, hsub hx.1⟩⟩
+
+
 
 end FBA
