--- a/VCV-io/VCVio/EvalDist/Basic.lean
+++ b/VCV-io/VCVio/EvalDist/Basic.lean
@@ -273,17 +273,16 @@
 
 @[simp] lemma evalDist_eq : evalDist p = liftM p := rfl
 
-@[simp] lemma probOutput_eq : probOutput p = p := by
-  refine funext fun x => ?_
-  simp only [probOutput_def, evalDist_eq, monad_pure_eq_pure, monad_bind_eq_bind, OptionT.run_mk,
-    pure_apply, Option.some.injEq, mul_ite, mul_one, mul_zero]
-  simp
-  refine (PMF.map_apply _ _ _).trans ?_
-  refine (tsum_eq_single x ?_).trans ?_
+theorem probOutput_eq: probOutput p = p := by
+  funext x
+  rw [probOutput_def, evalDist_eq, OptionT.run_monadLift]
+  change (PMF.map some p) (some x) = p x
+  rw [PMF.map_apply]
+  rw [tsum_eq_single x]
   · simp
-    refine fun x h h' => ?_
-    refine (h h'.symm).elim
-  simp only [↓reduceIte]
+  · intro b hb
+    simp [Option.some.injEq, hb.symm]
+
 
 @[simp] lemma probEvent_eq : probEvent p = p.toOuterMeasure := by
   refine funext fun x => ?_
