--- a/clean/Clean/Gadgets/Xor/Xor64.lean
+++ b/clean/Clean/Gadgets/Xor/Xor64.lean
@@ -114,14 +114,15 @@
   linarith [p_large_enough.elim]
 
 theorem completeness : Completeness (F p) elaborated Assumptions := by
-  intro i0 env input_var h_env input h_input as
-  let ⟨⟨ x0, x1, x2, x3, x4, x5, x6, x7 ⟩, ⟨ y0, y1, y2, y3, y4, y5, y6, y7 ⟩⟩ := input
+  intro i0 env input_var h_env input h_input has
+  rcases input with ⟨⟨x0, x1, x2, x3, x4, x5, x6, x7⟩, ⟨y0, y1, y2, y3, y4, y5, y6, y7⟩⟩
   simp only [circuit_norm, explicit_provable_type, Inputs.mk.injEq, U64.mk.injEq] at h_input
-  simp only [Assumptions, circuit_norm, U64.Normalized] at as
-  simp only [h_input, circuit_norm, main, ByteXorTable,
-    explicit_provable_type, Fin.forall_iff] at h_env ⊢
-  have h_env0 : env.get i0 = ↑(ZMod.val x0 ^^^ ZMod.val y0) := by simpa using h_env 0
+  simp only [Assumptions, circuit_norm, U64.Normalized] at has
+  simp only [h_input, circuit_norm, main, ByteXorTable, explicit_provable_type, Fin.forall_iff] at h_env ⊢
+  have h_env0 : env.get i0 = ↑(ZMod.val x0 ^^^ ZMod.val y0) := by
+    simpa using h_env 0
   simp_all [xor_val]
+
 
 def circuit : FormalCircuit (F p) Inputs U64 where
   Assumptions
