--- a/VCV-io/VCVio/OracleComp/OracleComp.lean
+++ b/VCV-io/VCVio/OracleComp/OracleComp.lean
@@ -476,10 +476,9 @@
   | pure x => simp [← h]
   | _ => simp at h
 
-@[simp]
-lemma pure_eq_bind_iff (oa : OracleComp spec α) (ob : α → OracleComp spec β) (y : β) :
-    pure y = oa >>= ob ↔ ∃ x : α, oa = pure x ∧ ob x = pure y :=
-  eq_comm.trans (bind_eq_pure_iff oa ob y)
+theorem pure_eq_bind_iff (oa : OracleComp spec α) (ob : α → OracleComp spec β) (y : β) : pure y = oa >>= ob ↔ ∃ x : α, oa = pure x ∧ ob x = pure y := by
+  exact eq_comm.trans (bind_eq_pure_iff oa ob y)
+
 
 alias ⟨_, bind_eq_pure⟩ := bind_eq_pure_iff
 alias ⟨_, pure_eq_bind⟩ := pure_eq_bind_iff
