--- a/loom/Loom/MonadAlgebras/WP/Basic.lean
+++ b/loom/Loom/MonadAlgebras/WP/Basic.lean
@@ -222,41 +222,44 @@
 
 variable [∀ α, CCPO (m α)] [MonoBind m]
 
-lemma repeat_inv (f : Unit -> β -> m (ForInStep β))
+theorem repeat_inv (f : Unit -> β -> m (ForInStep β))
   (inv : ForInStep β -> l) (measure : β -> Nat)
   init :
    (∀ b, triple (inv (.yield b)) (f () b) (fun | .yield b' => inv (.yield b') ⊓ ⌜ measure b' < measure b ⌝ | .done b' => inv (.done b'))) ->
    triple (inv (.yield init)) (Loop.forIn.loop f init) (fun b => inv (.done b)) := by
   intro hstep
-  have induc (C : β → Prop) (a : β) (h : ∀ x, (∀ y, measure y < measure x → C y) → C x): C a := by
-    have lem: ∀ n: Nat, ∀ x, measure x ≤ n →  C x := by
+  have induc (C : β → Prop) (a : β)
+      (h : ∀ x, (∀ y, measure y < measure x → C y) → C x) : C a := by
+    have lem : ∀ n : Nat, ∀ x, measure x ≤ n → C x := by
       intro n
       induction n with
       | zero =>
-        intro x hx
-        exact h x (fun y => by omega)
+          intro x hx
+          exact h x (fun y hy => by
+            omega)
       | succ m ih =>
-        intro x hx
-        by_cases neq: measure x ≤ m
-        { exact ih x neq }
-        have eq: measure x = m + 1 := by omega
-        exact h x (fun y hy => ih y (by omega))
+          intro x hx
+          by_cases hxm : measure x ≤ m
+          · exact ih x hxm
+          · have hxeq : measure x = m + 1 := by
+              omega
+            exact h x (fun y hy => ih y (by
+              omega))
     exact lem (measure a) a (by simp)
-  apply induc
-    (fun ini => triple (inv (.yield ini)) (Loop.forIn.loop f ( ini)) (fun b => inv (.done b))) init
-  intro b ih; unfold Loop.forIn.loop; simp [triple, wp_bind]; apply le_trans
-  apply hstep; apply wp_cons; rintro (_|_)
-  { simp [wp_pure] }
-  rename_i a
-  match ForInStep.yield a with
-  | .yield b' =>
-    simp; intro m_lt
-    have ihb := ih b' m_lt
-    simp [triple] at ihb
-    exact ihb
-  | .done b' =>
-    simp
-    rw [wp_pure b' (fun s: β => inv (ForInStep.done s))]
+  refine induc (fun ini => triple (inv (ForInStep.yield ini)) (Loop.forIn.loop f ini) (fun b => inv (ForInStep.done b))) init ?_
+  intro b ih
+  unfold Loop.forIn.loop
+  simp [triple, wp_bind]
+  apply le_trans
+  · exact hstep b
+  · apply wp_cons
+    rintro (_ | _)
+    · simp [wp_pure]
+    · simp
+      intro m_lt
+      have ihb := ih _ m_lt
+      simpa [triple] using ihb
+
 
 
 lemma repeat_inv_split (f : Unit -> β -> m (ForInStep β))
