--- a/iris-lean/src/Iris/Algebra/OFE.lean
+++ b/iris-lean/src/Iris/Algebra/OFE.lean
@@ -706,29 +706,36 @@
     | zero => exact H _
     | succ _ IH => exact (H _).trans <| Contractive.succ _ <| IH <| Dist.lt H (Nat.lt_add_one _)
 
-@[elab_as_elim]
-theorem OFE.ContractiveHom.fixpoint_ind [COFE α] [Inhabited α] (f : α -c> α)
-    (P : α → Prop) (HProper : ∀ A B : α, A ≡ B → P A → P B) (x : α) (Hbase : P x)
-    (Hind : ∀ x, P x → P (f x)) (Hlim : LimitPreserving P) :
-    P f.fixpoint := by
+theorem OFE.ContractiveHom.fixpoint_ind [COFE α] [Inhabited α] (f : α -c> α) (P : α → Prop) (HProper : ∀ A B : α, A ≡ B → P A → P B) (x : α) (Hbase : P x) (Hind : ∀ x, P x → P (f x)) (Hlim : LimitPreserving P) : P f.fixpoint := by
   let chain : Chain α := by
-    refine ⟨fun i => Nat.repeat f (i + 1) x, fun {n i} H => ?_⟩
+    refine ⟨fun i => Nat.repeat f.f (i + 1) x, fun {n i} H => ?_⟩
     induction n generalizing i with
-    | zero => simp [Nat.repeat]
+    | zero =>
+        simp [Nat.repeat]
     | succ _ IH =>
-      cases i <;> simp at H
-      exact Contractive.succ _ <| IH H
-  refine HProper _ _ (fixpoint_unique (x := COFE.compl chain) ?_) ?_
-  · refine equiv_dist.mpr fun n => ?_
+        cases i <;> simp at H
+        exact Contractive.succ _ <| IH H
+  have hfix : COFE.compl chain ≡ f.fixpoint := by
+    apply fixpoint_unique (x := COFE.compl chain)
+    refine equiv_dist.mpr fun n => ?_
     apply COFE.conv_compl.trans
-    refine .trans ?_ (f.ne.ne COFE.conv_compl).symm
-    induction n
-    · exact Contractive.zero f.f
-    · rename_i IH; apply Contractive.succ _ IH
-  · apply Hlim; intro n
+    refine .trans ?_ ((f.ne.ne COFE.conv_compl).symm)
     induction n with
-    | zero => exact Hind (Nat.repeat f.f 0 x) Hbase
-    | succ _ IH => apply Hind (Nat.repeat f.f _ x) IH
+    | zero =>
+        simpa [chain, Nat.repeat] using
+          (Contractive.zero f.f : f (Nat.repeat f.f 0 x) ≡{0}≡ f (f (Nat.repeat f.f 0 x)))
+    | succ n IH =>
+        simpa [chain, Nat.repeat] using (Contractive.succ f.f IH)
+  have hPchain : P (COFE.compl chain) := by
+    apply Hlim chain
+    intro n
+    induction n with
+    | zero =>
+        simpa [chain, Nat.repeat] using Hind (Nat.repeat f.f 0 x) Hbase
+    | succ n IH =>
+        simpa [chain, Nat.repeat] using Hind (Nat.repeat f.f (n + 1) x) IH
+  exact HProper _ _ hfix hPchain
+
 
 end Fixpoint
 
