--- a/veil/Veil/DSL/Action/Theory.lean
+++ b/veil/Veil/DSL/Action/Theory.lean
@@ -386,8 +386,9 @@
 
 theorem TwoState_sound'_ret_unit [LawfulAction act] (req : SProp σ) (ens : RProp σ PUnit) :
   act.triple req ens → act.toTwoState.triple req (ens () ·) := by
-  have heq : (ens () ·) = (∃ r, ens r ·) := by ext ; rw [exists_over_PUnit]
-  rw [heq] ; apply TwoState_sound'
+  intro h
+  simpa only [exists_over_PUnit] using (TwoState_sound' (act := act) req ens h)
+
 
 /-- This is used by `#recover_invariants_in_tr` in `Rabia.lean`. -/
 theorem TwoState_sound'_ret_unit' [LawfulAction act] {st : σ} (ens : RProp σ PUnit) :
