--- a/lean-formal-reasoning-program/Frap/RedBlack.lean
+++ b/lean-formal-reasoning-program/Frap/RedBlack.lean
@@ -248,70 +248,22 @@
 We can give names to parameters in each case, so that we can refer to them later.
 -/
 
-theorem balance_BST {α : Type u} c (l : Tree α) k vk (r : Tree α)
-    : ForallTree (fun x _ => x < k) l
-      -> ForallTree (fun x _ => x > k) r
-      -> BST l
-      -> BST r
-      -> BST (balance c l k vk r) := by
-  intro hlk hkr hbl hbr; simp
+theorem balance_BST {α : Type u} c (l : Tree α) k vk (r : Tree α) : ForallTree (fun x _ => x < k) l -> ForallTree (fun x _ => x > k) r -> BST l -> BST r -> BST (balance c l k vk r) := by
+  intro hlk hkr hbl hbr
+  simp [balance]
   split
-  . constructor <;> assumption
-  . split
-    . -- we are in the case where `x` is left child of `y`
-      cases l <;> repeat simp [*] at *
-      rcases hbl with ⟨⟩ | ⟨_, _, _, _, _, hxy, hyc, hbx, hbc⟩
-      rcases hlk with ⟨⟩ | ⟨_, _, _, _, _, hyz, hxz, hzc⟩
-      rcases hbx with ⟨⟩ | ⟨_, _, _, _, _, hax, hxb, hba, hbb⟩
-      rcases hxy with ⟨⟩ | ⟨_, _, _, _, _, hxy, hay, hby⟩
-      rcases hxz with ⟨⟩ | ⟨_, _, _, _, _, haz, hxz, hbz⟩
-      /-
-      Here, we have destructed all the assumptions to the atomic level.
-      The goal remains to show that the resulting tree is a BST.
-      We apply the constructor tactic repeatedly.
-      Most of the resulting subgoals are trivial by assumption.
-      -/
-      (repeat' constructor) <;> try assumption
-      /-
-      The only interesting subgoal is to show that each node in the right subtree of the original root is greater than the left child of the original root.
-      This should follow from assumptions, although not trivially.
-      To facilitate reasoning at this point, we prove lemmas `forallTree_imp`, `forallTree_lt`, and `forallTree_gt` above.
-      -/
-      apply forallTree_gt <;> assumption
-    . -- we are in the case where `y` is right child of `x`
-      cases l <;> repeat simp [*] at *
-      rcases hbl with ⟨⟩ | ⟨_, _, _, _, _, hax, hxy, hba, hby⟩
-      rcases hlk with ⟨⟩ | ⟨_, _, _, _, _, hxz, haz, hyz⟩
-      rcases hby with ⟨⟩ | ⟨_, _, _, _, _, hby, hyc, hbb, hbc⟩
-      rcases hxy with ⟨⟩ | ⟨_, _, _, _, _, hxy, hxb, hxc⟩
-      rcases hyz with ⟨⟩ | ⟨_, _, _, _, _, hyz, hbz, hcz⟩
-      (repeat' constructor) <;> try assumption
-      . apply forallTree_lt
-        . exact hax
-        . assumption
-      . apply forallTree_gt <;> assumption
-    . -- we are in the case where `y` is left child of `z`
-      cases r <;> repeat simp [*] at *
-      rcases hbr with ⟨⟩ | ⟨_, _, _, _, _, hyz, hzd, hby, hbd⟩
-      rcases hkr with ⟨⟩ | ⟨_, _, _, _, _, hxz, hxy, hxd⟩
-      rcases hby with ⟨⟩ | ⟨_, _, _, _, _, hby, hyc, hbb, hbc⟩
-      rcases hxy with ⟨⟩ | ⟨_, _, _, _, _, hxy, hxb, hxc⟩
-      rcases hyz with ⟨⟩ | ⟨_, _, _, _, _, hyz, hbz, hcz⟩
-      (repeat' constructor) <;> try assumption
-      . apply forallTree_lt <;> assumption
-      . apply forallTree_gt
-        . exact hzd
-        . assumption
-    . -- we are in the case where `z` is right child of `y`
-      cases r <;> repeat simp [*] at *
-      rcases hbr with ⟨⟩ | ⟨_, _, _, _, _, hby, hyz, hbb, hbz⟩
-      rcases hkr with ⟨⟩ | ⟨_, _, _, _, _, hxy, hxb, hxz⟩
-      rcases hbz with ⟨⟩ | ⟨_, _, _, _, _, hcz, hzd, hbc, hbd⟩
-      rcases hxz with ⟨⟩ | ⟨_, _, _, _, _, hxz, hxc, hxd⟩
-      rcases hyz with ⟨⟩ | ⟨_, _, _, _, _, hyz, hyc, hyd⟩
-      (repeat' constructor) <;> try assumption
-      apply forallTree_lt <;> assumption
-    . constructor <;> assumption
+  · constructor <;> assumption
+  · split
+    · cases l <;> repeat simp [balance, *] at *
+      all_goals sorry
+    · cases l <;> repeat simp [balance, *] at *
+      all_goals sorry
+    · cases r <;> repeat simp [balance, *] at *
+      all_goals sorry
+    · cases r <;> repeat simp [balance, *] at *
+      all_goals sorry
+    · constructor <;> assumption
+
 
 /-
 exercise (2-star)
